src/QueryRenderer.ts

```diff
@@ -482,8 +483,11 @@ class QueryRenderChild extends MarkdownRenderChild {
         timeUnit: unitOfTime.DurationConstructor = 'days',
         amount = 1,
     ) {
-        const errorMessage = '⚠️ Postponement requires a date: due or scheduled.';
-        if (!task.dueDate && !task.scheduledDate) return new Notice(errorMessage, 10000);
+        const errorMessage = '⚠️ Postponement requires a happens date: start, due or scheduled.';
+        if (task.happens.moment === null) {
```

Comment:
This use of `task.happens` is interesting...

1. At some point, a reminder date type will be added, and it will likely be added to `Task.happens`...
2. But the implementation of `Task.happens` is totally divorced from your new `HappensDate` type.

So:

1. now we have two different mechanisms to specify what goes into `happens` - with no link between them.
2. and this code is using one of each... which, if they are not changed at the same time, will give inconsistent results.

As such, this change creates a bit of a maintenance risk/pitfall, I feel.

Reply:
That sounds good.

Depending on what this PR ends up with, we also have:

- this developer-docs page, which could be used to note locations needing editing: [How do I add a new field to the Task class?](https://publish.obsidian.md/tasks-contributing/Code/How+do+I+add+a+new+field+to+the+Task+class)
- And the search-text `NEW_TASK_FIELD_EDIT_REQUIRED` to add to a comment in code, as a reminder

Reply:
ok, this is done and tested with the vault from the GitHub action

Reply:
> That sounds good.

Here's another idea... A couple of lines later, there is this code:

```ts
        const dateTypeToUpdate = getDateFieldToPostpone(task);
        if (dateTypeToUpdate === null) return;
```

Surely `getDateFieldToPostpone(task)` is doing the exact same thing as `!task.startDate && !task.dueDate && !task.scheduledDate` but has the advantage of already being covered by tests.

So we don't need to do that test in 2 different ways.

So maybe try this (untested):

```ts
        const dateTypeToUpdate = getDateFieldToPostpone(task);
        if (dateTypeToUpdate === null) {
            const errorMessage = '⚠️ Postponement requires a happens date: due, scheduled or start.';
            return new Notice(errorMessage, 10000);
        }
```

Which gives this diff:


```diff
Index: src/QueryRenderer.ts
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/src/QueryRenderer.ts b/src/QueryRenderer.ts
--- a/src/QueryRenderer.ts	(revision ba4a8a36b1f57d80aee69fd3d049ff9e991f339f)
+++ b/src/QueryRenderer.ts	(date 1701203650265)
@@ -483,14 +483,12 @@
         timeUnit: unitOfTime.DurationConstructor = 'days',
         amount = 1,
     ) {
-        const errorMessage = '⚠️ Postponement requires a happens date: due, scheduled or start.';
-        if (task.happens.moment === null) {
+        const dateTypeToUpdate = getDateFieldToPostpone(task);
+        if (dateTypeToUpdate === null) {
+            const errorMessage = '⚠️ Postponement requires a happens date: due, scheduled or start.';
             return new Notice(errorMessage, 10000);
         }
 
-        const dateTypeToUpdate = getDateFieldToPostpone(task);
-        if (dateTypeToUpdate === null) return;
-
         const { postponedDate, newTasks } = createPostponedTask(task, dateTypeToUpdate, timeUnit, amount);
 
         await replaceTaskWithTasks({
```

Reply:
Done

Reply:
And tested